### ISSUE TITLE
`xcvradm power` subcommand is wrong

### ISSUE THREAD

**Issue by bnaecker** [oxidecomputer/transceiver-control#79]
Doing some work on `madrid` I noticed that the `xcvradm power` subcommand is not correct. For example:

```
BRM44220001 # xcvradm -i sidecar0 -t1 power
Port  Power  Software-override
   1  Low    No
BRM44220001 # xcvradm -i sidecar0 -t2 power
Port  Power  Software-override
   2  High   Yes
BRM44220001 # xcvradm -i sidecar0 -t1,2 power
Port  Power  Software-override
   1  High   Yes
   2  High   No
```

We get different answers if we grab the power state individually than if we get it for multiple modules. I've looked at the messages that we're sending to the SP for these, and they look like what I expect. For example

```
BRM44220001 # xcvradm -i sidecar0 -t1,2 --log-level trace power > /dev/null
Dec 28 01:16:24.752 DEBG bound UDP socket, local_addr: Ok([::]:45299), interface: sidecar0
Dec 28 01:16:24.752 DEBG joined IPv6 multicast group, multicast_addr: ff02::1de:2
Dec 28 01:16:24.752 DEBG spawned IO task
Dec 28 01:16:24.752 TRCE received outgoing request, request: OutstandingHostRequest { request: HostRpcRequest { header: Header { version: 1, message_id: 0, message_kind: HostRequest }, message: Message { version: 1, body: HostRequest(Status(ModuleId(0x6))) }, data: None }, n_retries: 0, n_error_messages: 0, response_tx: Sender { inner: Some(Inner { state: State { is_complete: false, is_closed: false, is_rx_task_set: true, is_tx_task_set: false } }) } }, task: io
Dec 28 01:16:24.752 TRCE sent outgoing request, message: Message { version: 1, body: HostRequest(Status(ModuleId(0x6))) }, peer: [ff02::1de:2%80]:11112, task: io
Dec 28 01:16:24.756 TRCE packet received, peer: [fe80::aa40:25ff:fe05:600%80]:11112, n_bytes: 31, task: io
Dec 28 01:16:24.757 DEBG SP response message received, message: Message { version: 1, body: SpResponse(Status { modules: ModuleId(0x6), failed_modules: ModuleId(0x0) }) }, peer: [fe80::aa40:25ff:fe05:600%80]:11112, task: io
Dec 28 01:16:24.757 TRCE received trailing data, n_bytes: 2, data: Some([59, 59]), task: io
Dec 28 01:16:24.757 TRCE received outgoing request, request: OutstandingHostRequest { request: HostRpcRequest { header: Header { version: 1, message_id: 1, message_kind: HostRequest }, message: Message { version: 1, body: HostRequest(Read { modules: ModuleId(0x6), read: MemoryRead { page: Sff8636(Lower), offset: 0, len: 1 } }) }, data: None }, n_retries: 0, n_error_messages: 0, response_tx: Sender { inner: Some(Inner { state: State { is_complete: false, is_closed: false, is_rx_task_set: true, is_tx_task_set: false } }) } }, task: io
Dec 28 01:16:24.757 TRCE sent outgoing request, message: Message { version: 1, body: HostRequest(Read { modules: ModuleId(0x6), read: MemoryRead { page: Sff8636(Lower), offset: 0, len: 1 } }) }, peer: [ff02::1de:2%80]:11112, task: io
Dec 28 01:16:24.760 TRCE packet received, peer: [fe80::aa40:25ff:fe05:600%80]:11112, n_bytes: 35, task: io
Dec 28 01:16:24.760 DEBG SP response message received, message: Message { version: 1, body: SpResponse(Read { modules: ModuleId(0x6), failed_modules: ModuleId(0x0), read: MemoryRead { page: Sff8636(Lower), offset: 0, len: 1 } }) }, peer: [fe80::aa40:25ff:fe05:600%80]:11112, task: io
Dec 28 01:16:24.761 TRCE received trailing data, n_bytes: 2, data: Some([30, 17]), task: io
Dec 28 01:16:24.761 TRCE received outgoing request, request: OutstandingHostRequest { request: HostRpcRequest { header: Header { version: 1, message_id: 2, message_kind: HostRequest }, message: Message { version: 1, body: HostRequest(Read { modules: ModuleId(0x4), read: MemoryRead { page: Sff8636(Lower), offset: 93, len: 1 } }) }, data: None }, n_retries: 0, n_error_messages: 0, response_tx: Sender { inner: Some(Inner { state: State { is_complete: false, is_closed: false, is_rx_task_set: true, is_tx_task_set: false } }) } }, task: io
Dec 28 01:16:24.761 TRCE sent outgoing request, message: Message { version: 1, body: HostRequest(Read { modules: ModuleId(0x4), read: MemoryRead { page: Sff8636(Lower), offset: 93, len: 1 } }) }, peer: [ff02::1de:2%80]:11112, task: io
Dec 28 01:16:24.762 TRCE packet received, peer: [fe80::aa40:25ff:fe05:600%80]:11112, n_bytes: 34, task: io
Dec 28 01:16:24.762 DEBG SP response message received, message: Message { version: 1, body: SpResponse(Read { modules: ModuleId(0x4), failed_modules: ModuleId(0x0), read: MemoryRead { page: Sff8636(Lower), offset: 93, len: 1 } }) }, peer: [fe80::aa40:25ff:fe05:600%80]:11112, task: io
Dec 28 01:16:24.762 TRCE received trailing data, n_bytes: 1, data: Some([3]), task: io
Dec 28 01:16:24.762 TRCE received outgoing request, request: OutstandingHostRequest { request: HostRpcRequest { header: Header { version: 1, message_id: 3, message_kind: HostRequest }, message: Message { version: 1, body: HostRequest(Read { modules: ModuleId(0x2), read: MemoryRead { page: Cmis(Lower), offset: 26, len: 1 } }) }, data: None }, n_retries: 0, n_error_messages: 0, response_tx: Sender { inner: Some(Inner { state: State { is_complete: false, is_closed: false, is_rx_task_set: true, is_tx_task_set: false } }) } }, task: io
Dec 28 01:16:24.762 TRCE sent outgoing request, message: Message { version: 1, body: HostRequest(Read { modules: ModuleId(0x2), read: MemoryRead { page: Cmis(Lower), offset: 26, len: 1 } }) }, peer: [ff02::1de:2%80]:11112, task: io
Dec 28 01:16:24.764 TRCE packet received, peer: [fe80::aa40:25ff:fe05:600%80]:11112, n_bytes: 34, task: io
Dec 28 01:16:24.764 DEBG SP response message received, message: Message { version: 1, body: SpResponse(Read { modules: ModuleId(0x2), failed_modules: ModuleId(0x0), read: MemoryRead { page: Cmis(Lower), offset: 26, len: 1 } }) }, peer: [fe80::aa40:25ff:fe05:600%80]:11112, task: io
Dec 28 01:16:24.764 TRCE received trailing data, n_bytes: 1, data: Some([80]), task: io
BRM44220001 # xcvradm -i sidecar0 -t2 --log-level trace power > /dev/null
Dec 28 01:16:31.918 DEBG bound UDP socket, local_addr: Ok([::]:46303), interface: sidecar0
Dec 28 01:16:31.918 DEBG joined IPv6 multicast group, multicast_addr: ff02::1de:2
Dec 28 01:16:31.918 DEBG spawned IO task
Dec 28 01:16:31.918 TRCE received outgoing request, request: OutstandingHostRequest { request: HostRpcRequest { header: Header { version: 1, message_id: 0, message_kind: HostRequest }, message: Message { version: 1, body: HostRequest(Status(ModuleId(0x4))) }, data: None }, n_retries: 0, n_error_messages: 0, response_tx: Sender { inner: Some(Inner { state: State { is_complete: false, is_closed: false, is_rx_task_set: true, is_tx_task_set: false } }) } }, task: io
Dec 28 01:16:31.918 TRCE sent outgoing request, message: Message { version: 1, body: HostRequest(Status(ModuleId(0x4))) }, peer: [ff02::1de:2%80]:11112, task: io
Dec 28 01:16:31.922 TRCE packet received, peer: [fe80::aa40:25ff:fe05:600%80]:11112, n_bytes: 30, task: io
Dec 28 01:16:31.922 DEBG SP response message received, message: Message { version: 1, body: SpResponse(Status { modules: ModuleId(0x4), failed_modules: ModuleId(0x0) }) }, peer: [fe80::aa40:25ff:fe05:600%80]:11112, task: io
Dec 28 01:16:31.922 TRCE received trailing data, n_bytes: 1, data: Some([59]), task: io
Dec 28 01:16:31.922 TRCE received outgoing request, request: OutstandingHostRequest { request: HostRpcRequest { header: Header { version: 1, message_id: 1, message_kind: HostRequest }, message: Message { version: 1, body: HostRequest(Read { modules: ModuleId(0x4), read: MemoryRead { page: Sff8636(Lower), offset: 0, len: 1 } }) }, data: None }, n_retries: 0, n_error_messages: 0, response_tx: Sender { inner: Some(Inner { state: State { is_complete: false, is_closed: false, is_rx_task_set: true, is_tx_task_set: false } }) } }, task: io
Dec 28 01:16:31.922 TRCE sent outgoing request, message: Message { version: 1, body: HostRequest(Read { modules: ModuleId(0x4), read: MemoryRead { page: Sff8636(Lower), offset: 0, len: 1 } }) }, peer: [ff02::1de:2%80]:11112, task: io
Dec 28 01:16:31.926 TRCE packet received, peer: [fe80::aa40:25ff:fe05:600%80]:11112, n_bytes: 34, task: io
Dec 28 01:16:31.926 DEBG SP response message received, message: Message { version: 1, body: SpResponse(Read { modules: ModuleId(0x4), failed_modules: ModuleId(0x0), read: MemoryRead { page: Sff8636(Lower), offset: 0, len: 1 } }) }, peer: [fe80::aa40:25ff:fe05:600%80]:11112, task: io
Dec 28 01:16:31.926 TRCE received trailing data, n_bytes: 1, data: Some([17]), task: io
Dec 28 01:16:31.926 TRCE received outgoing request, request: OutstandingHostRequest { request: HostRpcRequest { header: Header { version: 1, message_id: 2, message_kind: HostRequest }, message: Message { version: 1, body: HostRequest(Read { modules: ModuleId(0x4), read: MemoryRead { page: Sff8636(Lower), offset: 93, len: 1 } }) }, data: None }, n_retries: 0, n_error_messages: 0, response_tx: Sender { inner: Some(Inner { state: State { is_complete: false, is_closed: false, is_rx_task_set: true, is_tx_task_set: false } }) } }, task: io
Dec 28 01:16:31.926 TRCE sent outgoing request, message: Message { version: 1, body: HostRequest(Read { modules: ModuleId(0x4), read: MemoryRead { page: Sff8636(Lower), offset: 93, len: 1 } }) }, peer: [ff02::1de:2%80]:11112, task: io
Dec 28 01:16:31.928 TRCE packet received, peer: [fe80::aa40:25ff:fe05:600%80]:11112, n_bytes: 34, task: io
Dec 28 01:16:31.928 DEBG SP response message received, message: Message { version: 1, body: SpResponse(Read { modules: ModuleId(0x4), failed_modules: ModuleId(0x0), read: MemoryRead { page: Sff8636(Lower), offset: 93, len: 1 } }) }, peer: [fe80::aa40:25ff:fe05:600%80]:11112, task: io
Dec 28 01:16:31.928 TRCE received trailing data, n_bytes: 1, data: Some([3]), task: io
```

So we're asking for the right information:

- Status for both modules
- Read the identifier for both modules
- Read the SFF-8636 module and then the CMIS (to fetch software-override / software power state)

It seems that the problem is interpretation of the data, likely something about how we're merging the hardware and software control.

**Comment by bnaecker** [comment]
Looks like this is a variant of #69. We're pushing all the parsed data into a vector in the order in which we read them, which may not be the same as the actual iteration order of the modules. I added some more debug logs, and you can see this:

```
BRM44220001 # /tmp/xcvradm -i sidecar0 -t 1 --log-level trace power > /dev/null
<SNIP>
Dec 28 23:36:37.400 TRCE read power control, control: ModuleResult { modules: ModuleId(0x2), data: [UseLpModePin], failures: FailedModules { modules: ModuleId(0x0), errors: [] } }
<SNIP>
BRM44220001 # /tmp/xcvradm -i sidecar0 -t 2 --log-level trace power > /dev/null
<SNIP>
Dec 28 23:36:51.415 TRCE read power control, control: ModuleResult { modules: ModuleId(0x4), data: [OverrideLpModePin { low_power: false }], failures: FailedModules { modules: ModuleId(0x0), errors: [] } }
<SNIP>
BRM44220001 # /tmp/xcvradm -i sidecar0 -t 1,2 --log-level trace power > /dev/null
<SNIP>
Dec 28 23:36:59.597 TRCE read power control, control: ModuleResult { modules: ModuleId(0x6), data: [OverrideLpModePin { low_power: false }, UseLpModePin], failures: FailedModules { modules: ModuleId(0x0), errors: [] } }
```

So reading from module 1 and 2 alone, we decode the power control as `UseLpModePin` and `OverrideLpModePin { low_power: false }`, respectively. That's correct. But when we read them together, we decode the array of control enums as `[OverrideLpModePin { low_power: false }, UseLpModePin]`, which is reversed. It should still be the case that the first module is `UseLpModePin`, and the second is the other. This can be traced to this line:

https://github.com/oxidecomputer/transceiver-control/blob/main/controller/src/controller.rs#L1143

We're reading each module, potentially out of order since we split the reads into all those with one management interface. (That's to take advantage of the parallelism and read from all the same _kind_ of modules in one message.) But then we iterate over them in the order in which they're read and push them into the vector. That needs to be ordered _by module index_ -- the easiest way to do that, since the reads may fail at any time, is a `BTreeMap` from index to the parsed data, which we then collect into a `Vec` at the end. Fix incoming.

**Comment by bnaecker** [comment]
Closed by #80
